pkg/guacanalytics/patchPlanning.go

```diff
@@ -0,0 +1,179 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package guacanalytics
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Khan/genqlient/graphql"
+	model "github.com/guacsec/guac/pkg/assembler/clients/generated"
+	"github.com/guacsec/guac/pkg/misc/depversion"
+)
+
+type DfsNode struct {
+	expanded bool // true once all node neighbors are added to queue
+	Parent   string
+	depth    int
+	nodeType string // packageName, packageVersion
+}
+
+type queueValues struct {
+	path    []string
+	nodeMap map[string]DfsNode
+	now     string
+	nowNode DfsNode
+	queue   []string
+}
+
+// TODO: make more robust using predicates
+func SearchDependenciesFromStartNode(ctx context.Context, gqlclient graphql.Client, startID string, stopID string, startType string, maxDepth int) (map[string]DfsNode, error) {
+	startNode, err := model.Node(ctx, gqlclient, startID)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed getting intial node with given ID:%v", err)
+	}
+
+	_, ok := startNode.Node.(*model.NodeNodePackage)
+
+	if !ok {
+		return nil, fmt.Errorf("Not a package")
+	}
+
+	q := queueValues{
+		queue:   make([]string, 0), // the queue of nodes in bfs
+		nodeMap: map[string]DfsNode{},
+	}
+
+	q.nodeMap[startID] = DfsNode{
+		nodeType: startType,
+	}
+
+	q.queue = append(q.queue, startID)
+
+	for len(q.queue) > 0 {
+		q.now = q.queue[0]
+		q.queue = q.queue[1:]
+		q.nowNode = q.nodeMap[q.now]
+
+		if stopID == q.now {
+			break
+		}
+
+		if q.nowNode.depth >= maxDepth {
+			break
+		}
+
+		neighborsResponse, err := model.Neighbors(ctx, gqlclient, q.now, []model.Edge{})
+
+		if err != nil {
+			return nil, fmt.Errorf("failed getting neighbors:%w", err)
+		}
+
+		for _, neighbor := range neighborsResponse.Neighbors {
+			q, err = caseOnPredicates(ctx, gqlclient, &q, neighbor)
+
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		q.nowNode.expanded = true
+		q.nodeMap[q.now] = q.nowNode
+	}
+
+	return q.nodeMap, nil
+
+}
+
+func caseOnPredicates(ctx context.Context, gqlclient graphql.Client, q *queueValues, neighbor model.NeighborsNeighborsNode) (queueValues, error) {
+	// case on predicates and nodeType
+	switch q.nowNode.nodeType {
+	case "packageVersion":
+		switch neighbor := neighbor.(type) {
+		case *model.NeighborsNeighborsIsDependency:
+			q, err := exploreIsDependency(ctx, gqlclient, q, *neighbor)
+
+			if err != nil {
+				return q, err
+			}
+		}
+	}
+
+	return *q, nil
+}
+
+func exploreIsDependency(ctx context.Context, gqlclient graphql.Client, q *queueValues, isDependency model.NeighborsNeighborsIsDependency) (queueValues, error) {
+	dependentPkgFilter := &model.PkgSpec{
+		Type:      &isDependency.DependentPackage.Type,
+		Namespace: &isDependency.DependentPackage.Namespaces[0].Namespace,
+		Name:      &isDependency.DependentPackage.Namespaces[0].Names[0].Name,
+	}
+
+	depPkgResponse, err := model.Packages(ctx, gqlclient, dependentPkgFilter)
```

Comment:
I think this is redundant since it should be addressable via `allIsDependencyTreeDependentPackage` embedding of NeighborsNeighborsIsDependency